### ISSUE TITLE
Piro: Epetra classes to check for int type before calling NumGlobalElements

### ISSUE THREAD

**Issue by bartgol** [trilinos/Trilinos#4651]
@trilinos/piro

## Expectations

One should be able to use Epetra maps with long long global indices, and still use Piro.

## Current Behavior

If the Epetra maps are built with long long global indices, piro epetra classes to not work, due to calls to `NumGlobalElements` rather than `NumGlobalElements64`.
 
## Motivation and Context

Albany needs to use 64bits global ordinals. Therefore, the epetra maps store long long global ordinals. However, the method `NumGlobalElements` in `Epetra_BlockMap` returns an int. One should use `NumGlobalElements64` to get a 64bits integer return value. Even more: the former method throws if the map actually stores 64bits integers.

## Definition of Done

The epetra classes in piro (e.g., Piro::Epetra::NOXSolver) should work with both 32 and 64 bits integers.

## Possible Solution

Before each call to NumGlobalElements, check for the type of indices. If they are int (32bits), use `NumGlobalElements`, while if they are long long (64bits), use `NumGlobalElements64`.

**Comment by mhoemmen** [comment]
@trilinos/piro 

I would be cautious about Epetra64.  Lots of things that work with 32-bit Epetra quite likely do not work with Epetra64, like AztecOO and ML.  MueLu probably does not test with Epetra64.

**Comment by bartgol** [comment]
Well, I guess Piro is one of those things, so far. Albany can easily switch back to 32 bits GO for epetra builds, of course. But if Epetra64 should not be used, then why bothering having it in the first place?

**Comment by mhoemmen** [comment]
If you need Epetra64 to work, your best bet would be to petition @jhux2 and the @trilinos/muelu team to support it fully.  Assuming that Epetra64 itself works, Belos should work with no or few changes.  Smoothers may or may not work, depending on the smoother.  AztecOO and ML definitely won't work.

I've worked on Tpetra for a long time, but I think I can recommend objectively that your time would be better spent converting to Tpetra.

> But if Epetra64 should not be used, then why bothering having it in the first place?

Epetra64 wasn't my idea.  (Trilinos developers had a debate about its necessity and wisdom at the time.) 
 Once we stood up Tpetra-based multigrid and demonstrated it at scale running real problems, users had less need for Epetra64.

**Comment by bartgol** [comment]
Ok, thanks for the insight. I was able to adapt Albany to use 32bits in case Epetra is used, to avoid trouble. I peeked inside Ifpack, ML, and AztecOO, and I saw plenty of places where `int` is assumet for the GIDs. At this point I think it's better to stay away from 64bits in Epetra, at least for our needs, and simply turn to Tpetra if that's needed.

I will close this issue, since it seems it goes beyond Piro, and is probably a wontfix thing.

@jhux2 no need to support Epetra64 in muelu on our end...yet. :)
